runtime/Cpp/runtime/src/ANTLRInputStream.cpp

```diff
@@ -17,44 +17,41 @@ using namespace antlrcpp;
 
 using misc::Interval;
 
+ANTLRInputStream::ANTLRInputStream() {
+  InitializeInstanceFields();
+}
+
 #if __cplusplus >= 201703L
-ANTLRInputStream::ANTLRInputStream(std::string_view input) {
-#else
-ANTLRInputStream::ANTLRInputStream(const std::string &input) {
+ANTLRInputStream::ANTLRInputStream(const std::string_view &input): ANTLRInputStream() {
+  load(input.data(), input.length());
+}
 #endif
-  InitializeInstanceFields();
-  load(input);
+
+ANTLRInputStream::ANTLRInputStream(const std::string &input): ANTLRInputStream() {
+  load(input.data(), input.size());
 }
 
-ANTLRInputStream::ANTLRInputStream(const char data_[], size_t numberOfActualCharsInArray)
-  : ANTLRInputStream(std::string(data_, numberOfActualCharsInArray)) {
+ANTLRInputStream::ANTLRInputStream(const char *data, size_t length) {
+  load(data, length);
 }
 
-ANTLRInputStream::ANTLRInputStream(std::istream &stream) {
-  InitializeInstanceFields();
+ANTLRInputStream::ANTLRInputStream(std::istream &stream): ANTLRInputStream() {
   load(stream);
 }
 
-#if __cplusplus >= 201703L
-void ANTLRInputStream::load(std::string_view input) {
-  // Remove the UTF-8 BOM if present.
-  constexpr std::string_view bom = "\xef\xbb\xbf";
-  if (input.compare(0, 3, bom) == 0)
-    input.remove_prefix(3);
-  _data = antlrcpp::utf8_to_utf32(input.data(), input.data() + input.size());
-  p = 0;
-}
-#else
 void ANTLRInputStream::load(const std::string &input) {
+  load(input.data(), input.size());
+}
+
+void ANTLRInputStream::load(const char *data, size_t length) {
   // Remove the UTF-8 BOM if present.
-  const char bom[4] = "\xef\xbb\xbf";
-  if (input.compare(0, 3, bom, 3) == 0)
-    _data = antlrcpp::utf8_to_utf32(input.data() + 3, input.data() + input.size());
+  const char *bom = "\xef\xbb\xbf";
+  if (length > 3 && strncmp(data, bom, 3) == 0)
```

Comment:
Maybe I'm just nitpicking at this point but wouldn't an empty bom be a valid empty string that could be valid for some grammars? So the check would be `length >= 3`. Or maybe I don't know my unicode. Sorry for all the disturbance.

Reply:
No worries, a good code review is often inconvenient. 

The approach there is that if there's enough to have a possible BOM then check it, if not just go ahead and load what you got. So even an empty string is "loaded" correctly in the else branch.

I'm assuming here that by "empty bom" you actually mean an empty input string.

Reply:
Here's what I mean: if only `"\xef\xbb\xbf"` is passed in the function, `length` will be 3 and the `else` branch will be taken. Then, at least when used with utfcpp conversion functions, it will throw `utf8::invalid_code_point` because it can't parse the bom.

Reply:
Yes, I also thought about this case and found it would not be worth to be considered as it would mean no useful input was given. However, there could be a workflow where a BOM is always attached automatically, regardless of what input was given and empty input is often valid input. So it would make sense to strip off the BOM and deal with the empty input anyway. I change the check therefore.